Initial changes for detecting unused imports.

m-kostrzewa opened the pull request:
Ran it on contrail-windows-ci repo, found a bunch of false positives: https://gist.github.com/m-kostrzewa/5bad3112320bf0c63d0012bc137c9567

False positives mainly for files that import files without defined functions (like Init.ps1) or with classes only (like Job.ps1).

Not sure how to handle either of those scenarios with current Shelly implementation, since it only handles functions for now - not classes or files without functions. Asking the user to specify `allow unused-imports` for all these files is too cumbersome IMO.

Suggested workaround: don't raise unused-imports, if file has no function definitions. This may create false negatives or whatever for unused classes.

@krdln thoughts?

m-kostrzewa committed:
Initial changes for detecting unused imports.

m-kostrzewa commented on src/preprocess.rs:
```diff
@@ -68,20 +68,19 @@ pub fn parse_and_preprocess(path: &Path, run_opt: &RunOpt, emitter: &mut Emitter
 /// Returns None if any of imports were not recognized
-fn resolve_imports(source_path: &Path, imports: Vec<syntax::Import>, emitter: &mut Emitter)
-    -> Result<Option<Vec<PathBuf>>, Error>
+fn resolve_imports(source_path: &Path, mut imports: Vec<syntax::Import>, emitter: &mut Emitter)
```
How can we avoid using mut here?

krdln replied:
```rust
let mut imports = imports;
```
in the first line.

Anyway, why do we want to remove this `mut`? (It's actually just a syntax sugar for this ↑ (in contrary to `&mut`, which changes a lot)).

Anyway... I think I don't like merging "import as a syntax element" and "import with resolved path" into one thing. We could add `preprocess::Import` as a wrapper of `syntax::Import`, but the simplest solution would be to just keep a list of resolved imports as `Map<PathBuf, syntax::Import>` (this would also remove the necessity of checking for `None` which I've seen a few times).

krdln commented on src/scope.rs:
```diff
@@ -110,6 +114,21 @@ pub fn analyze<'a>(files: &'a Map<PathBuf, Parsed>, config: &ConfigFile, emitter
+                    something_imported_is_used = true;
+                    break;
+                }
+            }
```
[`Iterator::any`](https://static.rust-lang.org/doc/master/std/iter/trat.Iterator.html#method.any) would be helpful here.

krdln replied:
You known I cry a little bit when I see nested `for` loops. If we apply my suggestion from above, we could just use a `used_import` set and be done with it.

krdln commented on src/scope.rs:
```diff
@@ -45,6 +46,9 @@ pub struct Scope<'a> {
     directly_imported_or_defined: Set<UniCase<&'a str>>,
+    /// Directly imported definitions grouped by file. This introduces denormalized
+    /// data between directly_imported and this field.
+    definitions_per_import: Map<&'a syntax::Import, Set<UniCase<&'a str>>>,
```
This struct is growing a little bit too... organically.

What do you think about instead of just having sets, keep a `Map` from name to "Item", where "Item" would contain info about the origin? (It could be a tuple, struct, or perhaps just `&Path`).

I think, this way we could get rid of all these `directly`/indirectly monstrocities (and just check whether the "origin" exists inside the direct-import list of the current file)

m-kostrzewa replied:
I was thinking about file graph here tbh

m-kostrzewa replied:
And then in scope.rs, instead of `for path in files.keys() {` we would DFS or whatever.

krdln replied:
Yup, a graph it could be (it's kind of a graph now, but anyway). But still, we need to have set of items in scope, preferably marked with their directness and origin and I think a simple `Map<Unicase<&str>, &Path>` would be a lot better than current "just make a 5 sets and a map and make sure you don't mess up your `extends`".

krdln commented on src/strictness.rs:
```diff
@@ -36,7 +36,12 @@ pub fn analyze<'a>(
+                Some(res) => {
+                    importees.insert(res);
+                }
+            }
```
You can use if-let instead of such a single-arm match.

krdln commented on src/syntax/mod.rs:
```diff
@@ -16,21 +17,33 @@ pub struct File {
+    fn partial_cmp(&self, other: &Import) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
```
If we treat the `Import` as a "syntactic element in the source code", ordering on just one field would be surprising. Perhaps it'd be nice to differentiate between "Import as a syntax" element and "Import as a path" (for the latter, I'd just use bare Path)

krdln commented:
Sorry for not grouping comments in one review, I thought I'd give just two quick comments and move on, but, you know...

> Not sure how to handle either of those scenarios with current Shelly implementation, since it only handles functions for now - not classes or files without functions. Asking the user to specify allow unused-imports for all these files is too cumbersome IMO.

Yup, that's why I wanted to finish the parser, so it gathers also classes and top-level variables before implementing this check.

> Suggested workaround: don't raise unused-imports, if file has no function definitions. This may create false negatives or whatever for unused classes.

It may even still create false positives – user imports a class+function file, but uses only a class.

I'd gate this lint, just as you described, but also, for now I'd set the default lint level to allow (to use you'd need too `shelly analyze --warn unused-lints`).

m-kostrzewa commented:
> I'd gate this lint

what do you mean by gate?

krdln commented:
> > I'd gate this lint, just as you described
>
> what do you mean by gate?

I meant to disable it on functionless files. Sorry for the wording.

krdln committed:
Move resolved_path from syntax::Import to map's key

krdln committed:
Merge branch 'master' into unused-import and break tests

I removed the Scope::definitions_per_import to simplify code,
but I've realized that it changes logic slightly.

krdln commented:
So... I tried to get rid of "definitions per import" and I've broken the tests.

So the problem is in combination of unused-imports with with indirect-imports. Your code took into account where the indirect import came from:

```rust
        let all_imported = nested.defined.union(&nested.directly_imported).cloned().collect();
        scope.definitions_per_import.insert(imported_file, all_imported);
```

whereas my reports "unused import" (because it's not used (at least directly)).
I think I will re-add something like your `definitions_per_import`.

Also, why `nested.defined ∪ nested.directly_imported`? That's not *all* imported.

krdln committed:
Check where indirect import came from

... so indirectly used imports are no longer marked as unused.
Also, add additional notes for indirect-imports lint,
so it's obvious how the item was imported and
where it was orignally defined.

krdln commented:
Ok, tests fixed. And I've improved indirect import diagnostics as a side effect :)

![indirect-import](https://user-images.githubusercontent.com/3074996/45850109-194a0d80-bd35-11e8-907f-5d294f55a0ae.png)

krdln committed:
Change unused-imports lint level to warn

Only code-breaking things should by default be red.

krdln committed:
Silence unused-imports on empty files

This is to avoid too many false-positives when importing
files without functions. There's still a risk of false
positive if one imports a file with both functions and
classes and uses only the classes, but let's keep this
lint on-by default for now and see how it performs.

krdln commented:
@m-kostrzewa LGTY?

Also, I'm not sure with warn-by-default. It produces quite a lot of warnings on contrail-windows-ci repo, even after implementing silencing we discussed. But I've looked into few of them and they seem legit.

Also, `clap` is awesome!
![clap-is-awesome-allow](https://user-images.githubusercontent.com/3074996/45851395-f7ec2000-bd3a-11e8-8d62-20610c6f0f94.png)

m-kostrzewa commented on src/scope.rs:
```diff
@@ -144,23 +137,72 @@ pub fn analyze<'a>(files: &'a Map<PathBuf, Parsed>, config: &ConfigFile, emitter
+                {
+                    // Temporarily silence unused-imports for files with no functions
+                    // to avoid false positives.
+                    // TODO Make the parser understand the world beyound functions
```
beyond*

m-kostrzewa commented on src/scope.rs:
```diff
@@ -138,23 +137,72 @@ pub fn analyze<'a>(files: &'a Map<PathBuf, Parsed>, config: &ConfigFile, emitter
+                if files[imported_file]
+                        .definitions
+                        .iter()
+                        .filter(|def| !def.name.starts_with("!")) // skip pseudoitems
```
I don't get this? What is a pseudoitem

krdln replied:
Pseudoitems are things not really defined in the source code, but something we inject in analysis. Eg. there's an pseudo-item ‘This file “defines” usage of strict mode’. Its name starts with `!` so it doesn't clash with real functions.

krdln replied:
Wrapped this iterator chain in a function, so it's more obvious now.

m-kostrzewa reviewed (commented):
lgtm

krdln committed:
Fix typo in comment

krdln committed:
Clarify skipping pseudoitems
